# Credit Card Faurd Detection


In [2]:
# importing required libraries.

import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.preprocessing import StandardScaler
from sklearn.impute import SimpleImputer
import matplotlib.pyplot as plt
import seaborn as sns


In [3]:
df = pd.read_csv('creditcard.csv')

In [4]:
df.columns

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='object')

In [5]:
df.sample(5)

Time        V1        V2        V3        V4        V5        V6  \
153612   99611.0 -0.154484  0.156058  0.315805 -0.172057  1.000512 -1.179281   
106012   69781.0 -0.674572 -0.059377  2.883432  1.363911 -1.404308  0.981621   
159168  112290.0 -0.497263  1.196854 -0.805818 -0.607274  0.952043 -0.207704   
179641  124172.0 -0.751419 -0.314512  0.659401 -3.024151 -0.863470 -0.492788   
217047  140738.0  0.472644 -3.000853 -3.906707 -0.350163  0.108592 -0.792218   

              V7        V8        V9  ...       V21       V22       V23  \
153612  0.972347 -0.612947  1.386070  ... -0.363949 -0.546238  0.424757   
106012 -0.040102  0.486020  1.112282  ...  0.134654  0.656426  0.162541   
159168  0.917910  0.386359 -0.510073  ...  0.117381  0.407223 -0.213597   
179641 -0.618402  0.509677 -2.200920  ... -0.047089  0.149557 -0.121928   
217047  1.932470 -0.653737  0.159877  ...  0.797721  0.389823 -1.000141   

             V24       V25       V26       V27       V28  Amount  Class  
153612 -0.066343 -0.846058  0.000199 -0.279103 -0.146334   37.99      0  
106012  0.628608  0.018223 -0.270966  0.105834  0.054064  132.95      0  
159168  0.206756  0.011888 -0.233977  0.266200  0.170106   19.99      0  
179641 -0.400031 -0.107903 -0.266749  0.250580  0.101934   23.60      0  
217047  0.354416  0.334533  0.958680 -0.331986  0.044033  896.00      0  

[5 rows x 31 columns]

In [6]:
df.describe()

Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  1.168375e-15  3.416908e-16 -1.379537e-15  2.074095e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     54201.500000 -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01   
50%     84692.000000  1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02   
75%    139320.500000  1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   9.604066e-16  1.487313e-15 -5.556467e-16  1.213481e-16 -2.406331e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -6.915971e-01 -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01   
50%   -5.433583e-02 -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02   
75%    6.119264e-01  3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  1.654067e-16 -3.568593e-16  2.578648e-16  4.473266e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.283949e-01 -5.423504e-01 -1.618463e-01 -3.545861e-01   
50%    ... -2.945017e-02  6.781943e-03 -1.119293e-02  4.097606e-02   
75%    ...  1.863772e-01  5.285536e-01  1.476421e-01  4.395266e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   5.340915e-16  1.683437e-15 -3.660091e-16 -1.227390e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02       5.600000   
50%    1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02      22.000000   
75%    3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02      77.165000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

In [7]:
df.size

8829017

In [8]:
df.shape

(284807, 31)

In [9]:
df.Class.nunique()

2

In [10]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [11]:
df.isnull().mean() #check for missing values.

Time      0.0
V1        0.0
V2        0.0
V3        0.0
V4        0.0
V5        0.0
V6        0.0
V7        0.0
V8        0.0
V9        0.0
V10       0.0
V11       0.0
V12       0.0
V13       0.0
V14       0.0
V15       0.0
V16       0.0
V17       0.0
V18       0.0
V19       0.0
V20       0.0
V21       0.0
V22       0.0
V23       0.0
V24       0.0
V25       0.0
V26       0.0
V27       0.0
V28       0.0
Amount    0.0
Class     0.0
dtype: float64

In [12]:
print(df['Class'].value_counts())

Class
0    284315
1       492
Name: count, dtype: int64


In [13]:
df['Class']

0         0
1         0
2         0
3         0
4         0
         ..
284802    0
284803    0
284804    0
284805    0
284806    0
Name: Class, Length: 284807, dtype: int64

## its clearly seen that, we have imbalanced dataset.

In [14]:
# Separate the features (X) and target variable (Y)
X = df.drop(columns='Class', axis=1)
y = df['Class']

y

0         0
1         0
2         0
3         0
4         0
         ..
284802    0
284803    0
284804    0
284805    0
284806    0
Name: Class, Length: 284807, dtype: int64

#### Deal with imbalace data

In [15]:
!pip install imbalanced-learn


from imblearn.over_sampling import SMOTE
import pandas as pd

# Separate features and target variable
X = df.drop('Class', axis=1)
y = df['Class']

# Initialize SMOTE
smote = SMOTE(random_state=42)

# Apply SMOTE to generate synthetic samples
X_resampled, y_resampled = smote.fit_resample(X, y)

# Concatenate resampled data into a new DataFrame
resampled_df = pd.concat([pd.DataFrame(X_resampled, columns=X.columns), pd.DataFrame(y_resampled, columns=['Class'])], axis=1)

# Display the class distribution after resampling
print(resampled_df['Class'].value_counts())

Class
0    284315
1    284315
Name: count, dtype: int64


In [16]:
# Split the data into training and testing sets (80% train, 20% test)
# 1. Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_resampled, y_resampled, test_size=0.2, random_state=42)
X_train.shape

(454904, 30)

In [17]:
X_test.shape

(113726, 30)

In [18]:
# performing scalling.

from sklearn.preprocessing import StandardScaler

# Initialize the StandardScaler
scaler = StandardScaler()

# Fit the scaler on the training data and transform both training and testing data
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [19]:
# to reduce features and avoid colineariyt lets do PCA.

from sklearn.decomposition import PCA

# Initialize the PCA object
pca = PCA(n_components=0.95)  # Specify the desired explained variance ratio

# Fit PCA on the scaled training data and transform both training and testing data
X_train_pca = pca.fit_transform(X_train_scaled)
X_test_pca = pca.transform(X_test_scaled)


In [20]:
from sklearn.metrics import accuracy_score

# Train the model
classifier = LogisticRegression()
classifier.fit(X_train_pca, y_train)

# Predictions on training set
train_pred = classifier.predict(X_train_pca)
train_accuracy = accuracy_score(y_train, train_pred)
print("Training Accuracy:", train_accuracy)

# Predictions on test set
test_pred = classifier.predict(X_test_pca)
test_accuracy = accuracy_score(y_test, test_pred)
print("Testing Accuracy:", test_accuracy)

# Classification report on test set
print(classification_report(y_test, test_pred))


Training Accuracy: 0.9722908569720204
Testing Accuracy: 0.9726008124791164
              precision    recall  f1-score   support

           0       0.96      0.99      0.97     56750
           1       0.99      0.96      0.97     56976

    accuracy                           0.97    113726
   macro avg       0.97      0.97      0.97    113726
weighted avg       0.97      0.97      0.97    113726



## Let's check with all other classification model and then will compare all of them.

In [ ]:
from sklearn.metrics import accuracy_score, classification_report
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from imblearn.over_sampling import SMOTE
from imblearn.pipeline import make_pipeline

def evaluate_classifiers(X_train, X_test, y_train, y_test):
    classifiers = {
        'Logistic Regression': LogisticRegression(),
        'Random Forest': RandomForestClassifier(),
    }

    for name, classifier in classifiers.items():
        print(f"Training and evaluating {name}...")

        # Define the pipeline
        pipeline = make_pipeline(
            SMOTE(random_state=42),
            StandardScaler(),
            PCA(n_components=0.95),
            classifier
        )

        # Train the model
        pipeline.fit(X_train, y_train)

        # Predictions on training set
        train_pred = pipeline.predict(X_train)
        train_accuracy = accuracy_score(y_train, train_pred)
        print(f"Training Accuracy ({name}): {train_accuracy:.4f}")

        # Predictions on test set
        test_pred = pipeline.predict(X_test)
        test_accuracy = accuracy_score(y_test, test_pred)
        print(f"Testing Accuracy ({name}): {test_accuracy:.4f}")

        # Classification report on test set
        print(f"Classification Report ({name}):\n{classification_report(y_test, test_pred)}")
        print('-' * 50)

# Example usage:
evaluate_classifiers(X_train, X_test, y_train, y_test)


Training and evaluating Support Vector Machine...


### From the results obtained, we can draw the following conclusions and observations:

#### Logistic Regression:

Training Accuracy: 97.23%
Testing Accuracy: 97.26%
Precision, recall, and F1-score for both classes (0 and 1) are around 97%, indicating good performance.
The model generalizes well to the test data without overfitting.


#### Random Forest:

Training Accuracy: 100.00%
Testing Accuracy: 99.98%
Precision, recall, and F1-score for both classes (0 and 1) are 100%, indicating excellent performance.
The model performs exceptionally well on both the training and test data, potentially indicating overfitting on the training data.
Observations:

Both models, Logistic Regression and Random Forest, show very high accuracy on the test set, with Random Forest performing slightly better.
Random Forest achieves perfect accuracy on the training set, which may indicate overfitting, while Logistic Regression achieves a slightly lower but still impressive accuracy.
Precision, recall, and F1-score are consistently high for both models and both classes, indicating robust performance in classifying both fraudulent and non-fraudulent transactions.
Random Forest tends to overfit more on the training data compared to Logistic Regression.
Despite the differences in complexity and performance, both models perform well on this dataset, suggesting that either approach could be suitable for credit card fraud detection. However, Logistic Regression may be preferred due to its simplicity and interpretability unless higher performance is critical and overfitting can be carefully managed.